add support for proxy_protocol for start_esmtp_listener

Solvik opened the pull request.

Solvik committed:
add support for proxy_protocol for start_esmtp_listener

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -650,13 +654,56 @@ impl EsmtpListenerParams {
+                                let peer_address = if proxy_protocol {
+                                    use tokio::io::AsyncReadExt;
+                                    let mut header_buf = [0u8; 16];
+                                    if let Err(e) = socket.read_exact(&mut header_buf).await {
```
I feel that an error reading the proxy packet should terminate the connection; if we fail to read then the connection is broken anyway.  If we fail to parse the proxy packet then we've already read and cannot un-read the data that should be interpreted differently, so we are guaranteed to have an SMTP protocol error as the next step anyway.

Solvik replied:
done!

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -650,13 +654,56 @@ impl EsmtpListenerParams {
+                                        orig_peer_address
+                                    } else {
+                                        // The length of the header is at bytes 14 and 15
+                                        let len = u16::from_be_bytes([header_buf[14], header_buf[15]]) as usize;
```
The proxy protocol stuff should really be factored out into a separate function rather than being placed inline here

Solvik replied:
done!

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -650,13 +654,56 @@ impl EsmtpListenerParams {
                             async move {
+                                let peer_address = if proxy_protocol {
+                                    use tokio::io::AsyncReadExt;
+                                    let mut header_buf = [0u8; 16];
```
the length of 16 is a magic number; presumably there is some constant in the `ppp` crate that we can use instead of encoding it here?

Solvik replied:
it's unfortunately not public, so I created a constant variable instead `PROXY_V2_MINIMUM_LENGTH`

wez reviewed (commented):
There should be a bit of design thinking in conjunction with future support for XCLIENT: specifically, we need to preserve the true remote ip address (of the proxy) in a context variable of some kind so that we don't lose that information for auditing purposes.

Precisely what that looks like depends on the goals of supporting the proxy protocol, so it's worth stating what you hope the outcome to be separately from the implementation in this PR.  I feel like it might be desirable to capture the proxy address in the trace headers somehow, but maybe not for everyone.  So it's worth you making a statement of your higher level goal: why do you want to support this and what specifically do you want it to enable?

There are also some security considerations; you'd want to only support this protocol for trusted connections, which implies that they must be connecting from a trust IP address, which implies that you should take care with how the listener definition is structured.  These are all things that should be mentioned in the docs that will also need to be supplied for this new option.

From a testing perspective, there should probably also be an integration test for this, which will require a bit of plumbing in the integration-test crate so that we have a source and a sink side that use the proxy and report the appropriate information and then assert that the appropriate context variables are set.

Solvik commented:
> There should be a bit of design thinking in conjunction with future support for XCLIENT: specifically, we need to preserve the true remote ip address (of the proxy) in a context variable of some kind so that we don't lose that information for auditing purposes.

I agree, we’ll preserve both addresses. My plan is to introduce:

```rust
pub enum ProxyMode { None, ProxyProtocol, XClient }

pub struct EsmtpListenerParams {
    // …
    pub proxy_mode: ProxyMode,
    pub trusted_proxy_hosts: Vec<SocketAddr>,
}
```

* peer_address remains the raw TCP socket peer (the load-balancer).
* client_address is either the same (when ProxyMode::None), or the real client IP decoded from PROXY protocol or XCLIENT.
* Both fields go into SmtpServerSession and into the Lua meta table.

> Precisely what that looks like depends on the goals of supporting the proxy protocol, so it's worth stating what you hope the outcome to be separately from the implementation in this PR. I feel like it might be desirable to capture the proxy address in the trace headers somehow, but maybe not for everyone. So it's worth you making a statement of your higher level goal: why do you want to support this and what specifically do you want it to enable?

When KumoMTA runs behind HAProxy, AWS NLB, or another TCP load-balancer, the server currently only sees the LB’s IP. Adding PROXY protocol now (and XCLIENT later) will let downstream Lua scripts see the true client IP for: auditing or extra Lua processing based on origin.

At the same time, we’ll still emit `peer_address` (the proxy’s socket) so you never lose that context.

> There are also some security considerations; you'd want to only support this protocol for trusted connections, which implies that they must be connecting from a trust IP address, which implies that you should take care with how the listener definition is structured. These are all things that should be mentioned in the docs that will also need to be supplied for this new option.

* We add a `trusted_proxy_hosts: Vec<SocketAddr>` (like `relay_hosts`).
* We only honor a PROXY or XCLIENT header if the raw TCP peer is in that list.
* If an untrusted host sends a header, we ignore it (and log a warning, or optionally drop).


> From a testing perspective, there should probably also be an integration test for this, which will require a bit of plumbing in the integration-test crate so that we have a source and a sink side that use the proxy and report the appropriate information and then assert that the appropriate context variables are set.

I’ll work on an integration test for PROXYv2 using ppp to craft packets.

Solvik committed:
- added a proxy_mode variable with an Enum, making xclient possible in a more generic way
- added a trusted_proxy_hosts variable
- switched relay_hosts check on client_address
- moved the proxy protocol parsing to a dedicated function

Solvik committed:
add timeout because we would wait indefinitly if there's no proxyv2 header

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -610,7 +627,22 @@ impl EsmtpListenerParams {
                     }
                     result = listener.accept() => {
-                        let (mut socket, peer_address) = result?;
+                        let (mut socket, orig_peer_address) = result?;
```
I think that we should move the handling of the proxy stuff into `SmtpServerSession::run` because it will simplify the configuration handling; right now the newly added proxy stuff is replicating (incompletely) the configuration resolution logic, and will be a source of bug reports if we don't address it!

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -650,13 +682,37 @@ impl EsmtpListenerParams {
+                                        Ok(Ok(addr)) => (orig_peer_address, addr),
+                                        Ok(Err(_)) | Err(_) => {
+                                            tracing::warn!("Proxy protocol v2 header invalid or timeout");
+                                            let response = format!("421 4.7.1 {} proxy protocol required\r\n", orig_peer_address);
```
SMTP requires that the hostname be included in a 421 response

```suggestion
                                            let response = format!("421 4.7.1 {hostname} {} proxy protocol required\r\n", orig_peer_address);
```

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -917,8 +978,8 @@ impl SmtpServerSession {
 
-    fn peer_in_cidr_list(&self, cidr: &CidrSet) -> bool {
-        cidr.contains(self.peer_address.ip())
+    fn client_in_cidr_list(&self, cidr: &CidrSet) -> bool {
```
I think this function should have been preserved as it was (which avoids the `unwrap`, which can in turn lead to a runtime panic, which is bad!) and instead, we should have `self.peer_address` be rewritten to be the peer on the other side of the proxy.

We should capture the address of the proxy as an optional metadata item.

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -853,13 +911,14 @@ impl SmtpServerSession {
         meta.set_meta("reception_protocol", "ESMTP");
         meta.set_meta("received_via", my_address.to_string());
-        meta.set_meta("received_from", peer_address.to_string());
+        meta.set_meta("client_address", client_address.to_string());
```
this breaks the `received_from` metadata and semantics; I think we should preserve this how it was.

Instead, when we handle the proxy protocol, we should capture the `peer_address` into a `proxy_address` (or better name) meta item, then update `received_from` to be the claimed peer address behind the proxy, as well as a `proxy_mode` (or better name) meta item describing what kind of proxy is in use.

I think that results in maximal sanity!

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -555,6 +566,11 @@ pub struct GenericEsmtpListenerParams {
     line_length_hard_limit: Option<usize>,
+
+    #[serde(default)]
+    pub proxy_mode: Option<ProxyMode>,
```
from a naming perspective, this might be better phrased as `allowed_proxy_mode` and possibly should be a list/set of `allowed_proxy_modes` so that we can subsequently support multiple options.

wez commented on crates/kumod/src/smtp_server.rs:
```diff
@@ -2764,3 +2825,50 @@ mod test {
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum ProxyMode {
+    None,
```
Let's remove the `None` variant from this, as we seem to be using `Option<ProxyMode>` throughout, and `Some(ProxyMode::None)` is redundant/ambiguous wrt. to `None`.

That will work better with using a list of allowed proxy modes; it would be better to have an empty list to indicate `no-allowed-proxy-modes` than a list containing `ProxyMode::None`, especially if that list could then contain some other proxy mode alongside it!

```suggestion
```

wez reviewed (commented):
Thanks; a few more comments here!

I think you can submit the next iteration to the main kumo repo, and that way I'll notice it and review it more easily and quickly!

wez commented:
one more thing to consider: after we've noticed that a proxy has been used, and updated the `peer_address`, we should re-consider the effective configuration by re-running the equivalent of:

```rust
concrete_params.apply_generic(params.base, &my_address, &peer_address, &mut meta);
```

with the updated peer address.

Solvik committed:
Update crates/kumod/src/smtp_server.rs

Co-authored-by: Wez Furlong <wez@wezfurlong.org>

wez commented:
FWIW, I just pushed https://github.com/KumoCorp/kumomta/commit/d9d3d14e3e541253be85854d59ac9a296c5d950b to enable XCLIENT.

This block of code is where the re-evaluation of the listener parameters is performed:

https://github.com/KumoCorp/kumomta/blob/d9d3d14e3e541253be85854d59ac9a296c5d950b/crates/kumod/src/smtp_server.rs#L2307-L2335
